Builder/AbstractFormContractor.php

```diff
@@ -0,0 +1,163 @@
+                    .' please review the UPGRADE-2.1.md file from the SonataAdminBundle',
+                    $type
+                ));
+            }
+            if (!$fieldDescription->getTargetEntity()) {
+                throw new \LogicException(sprintf(
+                    'The field "%s" in class "%s" does not have a target model defined.'
+                    .' Please make sure your association mapping is properly configured.',
+                    $fieldDescription->getName(),
+                    $fieldDescription->getAdmin()->getClass()
+                ));
+            }
+        }
+
+        if (in_array($type, array(
+            'sonata_type_admin',
+            'sonata_type_model_autocomplete',
+            'sonata_type_collection',
+        ), true)) {
+            if (!$fieldDescription->getAssociationAdmin()) {
+                throw $this->createMissingAssociationAdminException($fieldDescription);
+            }
+        }
+
+        if (in_array($type, array(
+            'sonata_type_model', 'sonata_type_model_list',
+            'sonata_type_model_hidden', 'sonata_type_model_autocomplete',
+        ), true)) {
+            $options['class'] = $fieldDescription->getTargetEntity();
+            $options['model_manager'] = $fieldDescription->getAdmin()->getModelManager();
+        } elseif ($type === 'sonata_type_admin') {
+            if (!$fieldDescription->describesSingleValuedAssociation()) {
+                throw new \LogicException(sprintf(
+                    'The `sonata_type_admin` type only handles single-valued associations.'
+                    .' Try using `sonata_type_collection` or one of the `sonata_type_model` variants'
+                    .' for field `%s`.',
+                    $fieldDescription->getName()
+                ));
+            }
+
+            // set sensitive default value to have a component working fine out of the box
+            $options['delete'] = false;
+
+            $options['data_class'] = $fieldDescription->getAssociationAdmin()->getClass();
+            $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'admin'));
+        } elseif ($type === 'sonata_type_collection') {
+            if (!$fieldDescription->describesCollectionValuedAssociation()) {
+                throw new \LogicException(sprintf(
+                    'The `sonata_type_collection` type only handles collection-valued associations.'
+                    .' Try using `sonata_type_admin` or one of the `sonata_type_model` variants'
+                    .' for field `%s`',
+                    $fieldDescription->getName()
+                ));
+            }
+            $options['type'] = 'sonata_type_admin';
+            $options['modifiable'] = true;
+            $options['type_options'] = array(
+                'sonata_field_description' => $fieldDescription,
+                'data_class' => $fieldDescription->getAssociationAdmin()->getClass(),
+            );
+        }
```

Comment:
@greg0ire Now that you point it, isn't `$fieldDescription->getAssociationAdmin()->getClass()` conceptually the same as `$fieldDescription->getTargetEntity()` ?